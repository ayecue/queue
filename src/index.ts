export { default as Queue, QueueOptions, QueueState } from './libs/queue';
export {
  default as Task,
  TaskCallback,
  TaskErrorCallback,
  TaskOptions,
  TaskState
} from './libs/task';
