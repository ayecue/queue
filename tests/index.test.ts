import { Queue } from '../src/index';
import { wait } from './utils';

describe('node-task-queue', function () {
  let testCallback;

  beforeEach(() => {
    testCallback = jest.fn();
  });

  describe('default', () => {
    test('normal queue with 1 concurrent', function (done) {
      const queue = new Queue({ concurrent: 1 });

      queue.enqueue({
        callback: async () => {
          await wait(500);
          testCallback(1);
        }
      });

      queue.enqueue({
        callback: async () => {
          await wait(500);
          testCallback(2);
        }
      });

      queue.enqueue({
        callback: async () => {
          await wait(500);
          testCallback(3);
        }
      });

      queue.on('idle', () => {
        expect(testCallback.mock.calls).toMatchSnapshot();
        done();
      });
    });

    test('normal queue with 2 concurrent', function (done) {
      const queue = new Queue({ concurrent: 2 });

      queue.enqueue({
        callback: async () => {
          await wait(500);
          testCallback(1);
        }
      });

      queue.enqueue({
        callback: async () => {
          await wait(500);
          testCallback(3);
        }
      });

      queue.enqueue({
        callback: async () => {
          await wait(500);
          testCallback(2);
        }
      });

      queue.on('idle', () => {
        expect(testCallback.mock.calls).toMatchSnapshot();
        done();
      });
    });

    test('normal queue with pause', function (done) {
      const queue = new Queue({ concurrent: 1 });

      queue.enqueue({
        callback: async () => {
          await wait(500);
          testCallback(1);
          queue.pause();
          wait(500).then(() => queue.unpause());
        }
      });

      queue.enqueue({
        callback: async () => {
          await wait(500);
          testCallback(2);
        }
      });

      queue.on('idle', () => {
        expect(testCallback.mock.calls).toMatchSnapshot();
        done();
      });
    });

    test('normal queue with cancel', function (done) {
      const queue = new Queue({ concurrent: 1 });

      queue.enqueue({
        callback: async () => {
          await wait(500);
          testCallback(1);
        }
      });

      const someTask = queue.enqueue({
        callback: async () => {
          await wait(500);
          testCallback(2);
        }
      });
      someTask.cancel();

      queue.enqueue({
        callback: async () => {
          await wait(500);
          testCallback(3);
        }
      });

      queue.on('idle', () => {
        expect(testCallback.mock.calls).toMatchSnapshot();
        done();
      });
    });
  });
});
