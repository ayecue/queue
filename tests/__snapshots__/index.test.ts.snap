// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`node-task-queue default normal queue with 1 concurrent 1`] = `
[
  [
    1,
  ],
  [
    2,
  ],
  [
    3,
  ],
]
`;

exports[`node-task-queue default normal queue with 2 concurrent 1`] = `
[
  [
    1,
  ],
  [
    3,
  ],
  [
    2,
  ],
]
`;

exports[`node-task-queue default normal queue with cancel 1`] = `
[
  [
    1,
  ],
  [
    3,
  ],
]
`;

exports[`node-task-queue default normal queue with pause 1`] = `
[
  [
    1,
  ],
  [
    2,
  ],
]
`;
